U2F support

jcj83429 opened the pull request:
This PR fulfills feature request #472.

We added support for U2F hardware keys as an alternative second factor. Users can choose between Google Authenticator and U2F when they signup. The feature can be enabled in the auth server configuration. 

This PR adds tstranex/u2f to the go dependencies and u2f-api-polyfill to the web dependencies.

We didn't update to the latest master because we had issues with bundled web assets (#619).

jcj83429 committed:
Implmented U2F registration and some of authentication on the server side

I know comments are very lacking right now. Once things are stable I will add
proper comments. Minimal manual testing of the U2F registration API was done
with a hardware U2F key. Some of the code may need to be cleaned up later to
remove excessively long variable names...

Currently we return an error rightaway if the username/password combo is wrong.
It's difficult to do U2F without revealing either whether a user exists or
whether the password is correct. Returning error immediately reveals whether
the user/password combo is valid, while waiting until we get a signed response
from the U2F device to announce whether the user/pass combo is valid can reveal
which users exist since we need to return a keyHandle in the U2F SignRequest
and generating fake keyHandles for nonexistent users can be difficult to get
right since there is no rigid format for keyHandle.

jcj83429 committed:
clean up unused imports due to removed test code

jcj83429 committed:
Implemented U2F authentication on the server side

jcj83429 committed:
add U2F Trusted Facets to configuration files. Trusted facets must include
the domain names of all proxies that users will log in with U2F from.

Commit:
make the config file entry names for u2f stuff consistent with what's
printed out by "teleport configure"

jcj83429 committed:
got u2f login working on the CLI client.

also grouped the u2f webapi endpoints together,
and fixed the default u2f AppID so it works out of the box

jcj83429 committed:
Merge upstream 'master' into u2f

Conflicts:
	lib/config/fileconf.go
	tool/tsh/main.go

jcj83429 committed:
make U2F AppId always lower case to match Chrome's behaviour

jcj83429 committed:
implemented web login with U2F

jcj83429 committed:
u2f signup working. UI needs polishing.

also made the secondFactorType var camel case everywhere

jcj83429 committed:
move u2f registration under the same path as other user credentials

jcj83429 committed:
move u2f sign challenges under the same path as other user credentials

jcj83429 committed:
no need to add new permission type for GetSignupU2fRegisterRequest

jcj83429 committed:
Merge branch 'master' into u2f

Conflicts:
	lib/config/fileconf.go
	web/src/app/services/auth.js

jcj83429 committed:
fix struct field tags causing test failure

jcj83429 committed:
add test for u2f-related persistence

also changed Get/UpsertU2fRegisterChallenge to return/take pointers
to be more consistent with the other Get/UpsertU2f* functions

jcj83429 committed:
added mock u2f device for testing

jcj83429 committed:
group the apiserver u2f endpoints together

jcj83429 committed:
document the u2f related web endpoints

bibekaur committed:
Fix css styling for 2fa options

jcj83429 committed:
fix U2fCRUD test

bibekaur committed:
Rewinding css changes made to login page

alchar committed:
refactored u2f login

alchar committed:
added u2f login tests

alchar committed:
fixed 2nd factor login with GoogleAuth hang

alchar committed:
fixed GoogleAuth bug

jcj83429 committed:
fix u2f login javascript refactor

jcj83429 committed:
refactor u2fSignUp to use promises.

alchar committed:
added u2f login tests, did some minor cleanup

jdahiya committed:
added tests for U2F sign up and tests for when Upserting Passwords using U2F is added

alchar committed:
Merge branch 'u2f-web-testing' of https://github.com/jcj83429/teleport into u2f-web-testing

alchar committed:
merged u2f-web-testing

jcj83429 committed:
add u2f appid to the config.js.

will make web ui hide u2f-related stuff if u2f appid is empty

jcj83429 committed:
hide u2f related stuff from web ui if appid is set to empty

to allow the admin to disable u2f

jdahiya committed:
Cleaned up redundencies in the test for U2F

jdahiya committed:
Added GetU2FSignRequest function to AuthWithRoles for dependency in ClientI implementation

jcj83429 committed:
complete rename of U2fSignRequest to GetU2fSignRequest

jcj83429 committed:
cleanup unused constant, duplicate code, and commented out code

jcj83429 committed:
add tests for u2f login api

jcj83429 committed:
Merge branch 'master' into u2f

jcj83429 committed:
add documentation on u2f usage

jcj83429 committed:
add u2f-api-polyfill to web/package.json

jcj83429 committed:
vendor github.com/tstranex/u2f library

jcj83429 committed:
rename u2f apis to remove underscores

gravitational-jenkins commented:
Can one of the admins verify this patch?

kontsevoy commented:
@jcj83429 can you guys resolve conflicts please? Thanks!

kontsevoy commented:
@jcj83429 see https://github.com/gravitational/teleport/blob/master/README.md#rebuilding-web-ui on how to rebuild the web UI.

jdahiya committed:
Fixed Format for Merging

jdahiya committed:
Fixed merge conflicts with original repository

jdahiya committed:
Fixed placement of the 'Insert U2f key' text to be below all buttons

jdahiya committed:
Added additional documentation

jdahiya committed:
Fixed Documentation to be strictly relevant to U2F implementation

jdahiya committed:
Fixed U2F CLI usage documentation

jdahiya commented:
@kontsevoy We've resolved all conflicts

klizhentas commented:
ok to test

jdahiya commented:
@kontsevoy @klizhentas Would it be possible to have the details as to where/why our build failed?

kontsevoy commented:
@jdahiya:

```
----------------------------------------------------------------------
FAIL: services_test.go:106: BoltSuite.TestU2fCRUD

services_test.go:107:
    s.suite.U2fCRUD(c)
/gopath/src/github.com/gravitational/teleport/lib/services/suite/suite.go:435:
    c.Assert(*challenge, DeepEquals, *challengeOut)
... obtained u2f.Challenge = u2f.Challenge{Challenge:[]uint8{0xc1, 0x41, 0xc8, 0xa4, 0x75, 0xa0, 0x3e, 0xdb, 0x38, 0x1b, 0xa1, 0xad, 0x68, 0xa7, 0xb6, 0xe4, 0xc1, 0xb2, 0x2b, 0x83, 0xd6, 0x5b, 0xe9, 0x68, 0x72, 0x8a, 0x3b, 0xdd, 0x10, 0x6f, 0x64, 0x9b}, Timestamp:time.Time{sec:63616165700, nsec:828887945, loc:(*time.Location)(0xec3f20)}, AppID:"https://localhost", TrustedFacets:[]string{"https://localhost"}}
... expected u2f.Challenge = u2f.Challenge{Challenge:[]uint8{0xc1, 0x41, 0xc8, 0xa4, 0x75, 0xa0, 0x3e, 0xdb, 0x38, 0x1b, 0xa1, 0xad, 0x68, 0xa7, 0xb6, 0xe4, 0xc1, 0xb2, 0x2b, 0x83, 0xd6, 0x5b, 0xe9, 0x68, 0x72, 0x8a, 0x3b, 0xdd, 0x10, 0x6f, 0x64, 0x9b}, Timestamp:time.Time{sec:63616165700, nsec:828887945, loc:(*time.Location)(0xe9d700)}, AppID:"https://localhost", TrustedFacets:[]string{"https://localhost"}}
```

jcj83429 committed:
Avoid using DeepEquals to compare time

to avoid test failures caused by Time.Location (time zone) differences.

jdahiya commented:
@kontsevoy Thank you, we've resolved the issue.

jdahiya committed:
Merge branch 'master' into u2f

klizhentas commented:
Hey All,

First of all, thanks for your PR - this is awesome. 

Second, I wanted to let you know that review is in progress, and we will be posting updates on it as it goes, however please be ready for longer collaboration on this PR given it's size and scope.

Thanks,
Sasha

jdahiya committed:
Merge branch 'master' into u2f

klizhentas commented on lib/services/identity.go:
```diff
@@ -199,6 +201,30 @@ type Identity interface {
 	DeleteSignupToken(token string) error
 
+	// upserts a U2F challenge for a new user corresponding to the token
+	UpsertU2fRegisterChallenge(token string, u2fChallenge *u2f.Challenge) (error)
```
let's rename those as well

klizhentas commented on lib/services/identity.go:
```diff
@@ -199,6 +201,30 @@ type Identity interface {
+	// upserts a U2F challenge for a new user corresponding to the token
+	UpsertU2fRegisterChallenge(token string, u2fChallenge *u2f.Challenge) (error)
+
+	// returns a U2F challenge for a new user corresponding to the token
```
go comments begin with the function names according to this guideline that we follow

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

klizhentas commented on lib/services/local/users.go:
```diff
@@ -419,6 +425,156 @@ func (s *IdentityService) DeleteSignupToken(token string) error {
 }
 
+// This is hardcoded in the U2F library
+const u2fChallengeTimeout = 5 * time.Minute
```
lets move this to defaults:

https://github.com/gravitational/teleport/blob/master/lib/defaults/defaults.go

klizhentas commented on lib/web/sessions.go:
```diff
@@ -228,6 +230,45 @@ func (s *sessionCache) Auth(user, pass string, hotpToken string) (*auth.Session,
+	return u2fSignReq, nil
+}
+
+func (s *sessionCache) AuthWithU2fSignResponse(user string, u2fSignResponse *u2f.SignResponse) (*auth.Session, error) {
```
could be simply `response`

klizhentas commented on lib/web/sshlogin.go:
```diff
@@ -196,6 +201,89 @@ func SSHAgentLogin(proxyAddr, user, password, hotpToken string, pubKey []byte, t
+	}
+
+	// Pass the JSON-encoded data undecoded to the u2f-host binary
+	u2fFacet := "https://" + strings.ToLower(proxyAddr)
```
simply `facet`

klizhentas commented on lib/web/sshlogin.go:
```diff
@@ -196,6 +201,89 @@ func SSHAgentLogin(proxyAddr, user, password, hotpToken string, pubKey []byte, t
+	signResponseLen := 500 + len(u2fSignRequest.Bytes()) + len(proxyAddr) * 4 / 3
+	signResponseBuf := make([]byte, signResponseLen)
+	signResponseLen, err = io.ReadFull(stdout, signResponseBuf)
+	// unexpected EOF means we have read the data completely.
```
according to the docs of  https://golang.org/pkg/io/#ReadFull

```
ReadFull reads exactly len(buf) bytes from r into buf. It returns the number of bytes copied and an error if fewer bytes were read. The error is EOF only if no bytes were read. If an EOF happens after reading some but not all the bytes, ReadFull returns ErrUnexpectedEOF. On return, n == len(buf) if and only if err == nil.
```

so `EOF` means no data, `ErrUnexpectedEOF` means not all data read, and `err==nil` means we have just fit in, so if I read this right we need to check for `EOF` 

```go
if err != nil || err != io.ErrUnexpectedEOF {
   return nil, trace.LimitExceeded("u2f sign response exceeded buffer size")
}
```

jcj83429 replied:
My understanding is that UnexpectedEOF means we read all data before filling the buffer, which is what we want. No error would mean there can still be more bytes in the pipe.

We don't check for EOF because the u2f-host program returns nothing in stdout when there is an error (and nothing in stderr when the sign was successful). We want to wait until we get the return code then decide if we should take the error message from stderr or the sign response from stdout. That said, we added an additional check after checking the return code to make sure we have read something from stdout in the case that u2f-host returned successfully.

klizhentas commented on lib/web/sshlogin.go:
```diff
@@ -196,6 +201,89 @@ func SSHAgentLogin(proxyAddr, user, password, hotpToken string, pubKey []byte, t
+
+	// Read error message (if any). 100 bytes is more than enough for any error message u2f-host outputs
+	errMsgBuf := make([]byte, 100)
+	errMsgLen, _ := io.ReadFull(stderr, errMsgBuf)
```
let's check error

klizhentas reviewed (changes requested):
Hey all, I've done the first part of the review. Implementation looks good from my side, however I have a bunch of style/code org comments.

What to expect next:

* UI part review from @alex-kovoy 
* I will test it with YubiKeys this week and will make one more pass to see if I've missed anything security wise.

So looks like we can get it in early next week assuming you can apply all code review comments from this week!

jcj83429 committed:
rename web APIs

signinpreauth -> signin/preauth
u2f/inviteregisterrequest -> u2f/signuptokens
u2f/newuser -> u2f/users
u2f/signrequest/:user -> u2f/users/:user/sign

jcj83429 committed:
group the u2f configs in auth_service

OLD:
auth_service:
  u2fappid: https://mycorp.com/appid.js
  u2ftrustedfacets:
  - https://proxy1.mycorp.com:3080
  - https://proxy2.mycorp.com:3080

NEW:
auth_service:
  u2f:
    enabled: yes
    appid: https://mycorp.com/appid.js
    facets:
    - https://proxy1.mycorp.com:3080
    - https://proxy2.mycorp.com:3080

alchar commented on lib/auth/apiserver.go:
```diff
@@ -572,6 +604,15 @@ func (s *APIServer) getDomainName(w http.ResponseWriter, r *http.Request, p http
 
+// getU2fAppId returns the U2F AppID in the auth configuration
+func (s *APIServer) getU2fAppId(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	appid, err := s.a.GetU2fAppId()
```
Should we change cases where u2f is the beginning of a variable name?

For example, change u2fReg to U2FReg

alchar committed:
Some cleanup for PR. Mostly appId -> appID, U2f -> U2F

alchar committed:
mocku2f uses trace for err handling

alchar committed:
missed one name change

jcj83429 committed:
fix sshlogin error handling

jcj83429 committed:
new_web_user.go: avoid repeatedly constructing lock path

jcj83429 committed:
new_web_user.go: rename u2fRegReq -> request

jcj83429 committed:
new_web_user.go: use trace.DebugReport to log error

jcj83429 committed:
revendor gravitational/trace to get trace.DebugReport

jcj83429 committed:
add comments to U2F related functions in tun.go

jcj83429 committed:
mocku2f: add support for different key handles

and general code cleanup

jcj83429 commented on lib/auth/mocku2f/mocku2f.go:
```diff
@@ -0,0 +1,173 @@
+		Subject: pkix.Name{
+			Organization: []string{"Test CA"},
+		},
+		NotBefore: time.Now(),
```
I think that won't be very useful. Currently we are skipping the certificate check during registration. The reason we still generate a self-signed certificate here is because the message format requires it. If we do enable certificate verification, the self-signed certificate won't verify anyway. The best we can do here is make the time range longer.

jcj83429 committed:
fix test failure caused by updating trace

jcj83429 committed:
change fileconf key appid to app_id

jcj83429 committed:
deduce U2F AppID from hostname if AppID not specified

jcj83429 committed:
Update admin guide to reflect changes in config file format

jcj83429 committed:
fix hiding of u2f-related things in web ui when u2f is disabled

jcj83429 committed:
Merge branch 'master' into u2f-devel

jcj83429 commented:
@klizhentas Thanks for the comments. We have implemented the suggested changes.

klizhentas commented:
retest this please

klizhentas commented:
hey all, I will test it on Monday once the YubiKeys get delivered, and will make a second pass as well.

@alex-kovoy do you have any feedback on the UI changes?

kontsevoy commented:
Woo! Nice. Thank you @jcj83429 +team!
